config.txt
!mode=test
!save_dir=nn/Pygeon/models
!data_dir=nn/Pygeon/data

#LeNet5
#pretrained=pretrained_models/LeNet5/LeNet5_MNIST_standard_best.bin
#10000: 0.93 (Frac 5)
#pretrained=pretrained_models/LeNet5/LeNet5_MNIST_custom_best.bin
#10000: 1.22 (Frac 5)
#pretrained=pretrained_models/old_models/lenet_temp.bin
#does not match
#pretrained=pretrained_models/old_models_hpmpc/lenet.bin
#does not match
#pretrained=pretrained_models/old_models_hpmpc/lenet5.pth
#does not match
#pretrained=pretrained_models/old_models_hpmpc/lenet5_avg.pth
#10000: 13.86 (Frac 5)
#pretrained=pretrained_models/old_models_hpmpc/lenet5_batch_avg.pth
#pretrained=pretrained_models/old_models_hpmpc/lenet5_batch_max_20.pth
#pretrained=pretrained_models/old_models_hpmpc/lenet_no_batch.pth
#10000: 13.86 (Frac 5)
#pretrained=pretrained_models/old_models_hpmpc/lenet_torch.bin
#does not match

#AlexNet32
#pretrained=pretrained_models/old_models_hpmpc/alex32.bin
#100 79 (Frac 5)
#pretrained=pretrained_models/old_models/alexnet32t_cifar10_temp.bin
#100 88 (Frac 5)
#pretrained=pretrained_models/old_models/og_alexnet32t_cifar10.bin
#100 82 (Frac 5)
#pretrained=pretrained_models/old_models_hpmpc/alex32.bin
#100 79 (Frac 5)
#pretrained=pretrained_models/cifar10_ADAM_001/AlexNet_32_CIFAR-10_standard_best.bin
#100 71 (Frac 5)

#pretrained=pretrained_models/old_models/AlexNet32_avg_c10_custom_best_50.bin
#100 76 (Frac 5)
#pretrained=pretrained_models/cifar10_ADAM_001/AlexNet_32_CIFAR-10_custom_best.bin
#100 63 (Frac 5)
#pretrained=pretrained_models/cifar10_ADAM_005/AlexNet_32_CIFAR-10_standard_best.bin
#100 62 (Frac 5)
#pretrained=pretrained_models/cifar10_ADAM_005/AlexNet_32_CIFAR-10_custom_best.bin
#100 68 (Frac 5)
#pretrained=pretrained_models/cifar10_SGD_001/AlexNet_32_CIFAR-10_standard_best.bin
#100 63 (Frac 5)
$pretrained=pretrained_models/cifar10_SGD_001/AlexNet_32_CIFAR-10_custom_best.bin
#100 70 (Frac 5)


#AlexNetC

#pretrained=pretrained_models/cifar10_ADAM_001/AlexNet_CryptGPU_CIFAR-10_standard_best.bin
#100 29 (Frac 5)
#pretrained=pretrained_models/cifar10_ADAM_005/AlexNet_CryptGPU_CIFAR-10_standard_best.bin
#100 25 (Frac 5)
#pretrained=pretrained_models/cifar10_SGD_001/AlexNet_CryptGPU_CIFAR-10_standard_best.bin
#100 28 (Frac 5)


#pretrained=pretrained_models/old_models/AlexNetC_avg_c10_custom_best_50.bin
#100 26 (Frac 5)
#pretrained=pretrained_models/cifar10_ADAM_001/AlexNet_CryptGPU_CIFAR-10_custom_best.bin
#100 26 (Frac 5)
#pretrained=pretrained_models/cifar10_ADAM_005/AlexNet_CryptGPU_CIFAR-10_custom_best.bin
#100 43 (Frac 5)
#pretrained=pretrained_models/cifar10_SGD_001/AlexNet_CryptGPU_CIFAR-10_custom_best.bin
#100 39 (Frac 5)

#VGG16

#pretrained=pretrained_models/old_models_hpmpc/vgg16_cifar.bin
#100 31 (Frac 5)
#pretrained=pretrained_models/old_models/og_vgg16_cifar10.bin
#100 93 (Frac 5)
#pretrained=pretrained_models/old_models/og_vgg16_cifar10B.bin
#does not match
#pretrained=pretrained_models/old_models/vgg_c10_standard_100_best.bin
#100 93 (Frac 5)
#pretrained=pretrained_models/cifar10_ADAM_001/VGG16_CIFAR-10_standard_best.bin
#100 92 (Frac 5)
#pretrained=pretrained_models/cifar10_ADAM_005/VGG16_CIFAR-10_standard_best.bin
#100 90 (Frac 5)
#pretrained=pretrained_models/cifar10_SGD_001/VGG16_CIFAR-10_standard_best.bin
#100 88 (Frac 5)

#pretrained=pretrained_models/cifar10_ADAM_001/VGG16_CIFAR-10_custom_best.bin
#100 93 (Frac 5)
#pretrained=pretrained_models/cifar10_ADAM_005/VGG16_CIFAR-10_custom_best.bin
#100 89 (Frac 5)
#pretrained=pretrained_models/cifar10_SGD_001/VGG16_CIFAR-10_custom_best.bin
#100 88 (Frac 5)

ResNet18

#pretrained=pretrained_models/cifar10_ADAM_001/ResNet18_avg_CIFAR-10_standard_best.bin
#100 39 (Frac 5)
#pretrained=pretrained_models/cifar10_ADAM_005/ResNet18_avg_CIFAR-10_standard_best.bin
#100 36 (Frac 5)
#pretrained=pretrained_models/cifar10_SGD_001/ResNet18_avg_CIFAR-10_standard_best.bin
#100 40 (Frac 5)

#pretrained=pretrained_models/cifar10_ADAM_001/ResNet18_avg_CIFAR-10_custom_best.bin
#100 40 (Frac 5)
#pretrained=pretrained_models/cifar10_ADAM_005/ResNet18_avg_CIFAR-10_custom_best.bin
#100 49 (Frac 5)
#pretrained=pretrained_models/cifar10_SGD_001/ResNet18_CIFAR-10_custom_best.bin
#100 54 (Frac 5)


#ResNet50

#pretrained=pretrained_models/old_models_hpmpc/ResNet50_cifar.bin
#100 78 (Frac 5)
#pretrained=pretrained_models/old_models_hpmpc/Resnet50_cifar.bin
#100 78 (Frac 5)
#pretrained=pretrained_models/cifar10_ADAM_001/ResNet50_avg_CIFAR-10_standard_best.bin
#100 56 (Frac 5)
#pretrained=pretrained_models/cifar10_ADAM_005/ResNet50_avg_CIFAR-10_standard_best.bin
#100 55 (Frac 5)
#pretrained=pretrained_models/cifar10_SGD_001/ResNet50_avg_CIFAR-10_standard_best.bin
#100 42 (Frac 5)

#pretrained=pretrained_models/old_models_hpmpc/ResNet50_custom.bin
#100 82 (Frac 5)
#pretrained=pretrained_models/old_models_hpmpc/ResNet50_custom2.bin
#100 74 (Frac 5)
#pretrained=pretrained_models/old_models/ResNet50_avg_c10_custom_best_50.bin
#100 38 (Frac 5)
#pretrained=pretrained_models/cifar10_ADAM_001/ResNet50_avg_CIFAR-10_custom_best.bin
#100 83 (Frac 5)
#pretrained=pretrained_models/cifar10_ADAM_005/ResNet50_avg_CIFAR-10_custom_best.bin
#100 88 (Frac 5)
#pretrained=pretrained_models/cifar10_SGD_001/ResNet50_avg_CIFAR-10_custom_best.bin
#100 80 (Frac 5)

#pretrained=pretrained_models/old_models_hpmpc/resnet50_cifar100.bin
#does not match
#pretrained=pretrained_models/old_models_hpmpc/ResNet50_Cifar_Max.bin
#100 78 (Frac 5) (avg instead of max)
#pretrained=pretrained_models/old_models_hpmpc/Restnet50_Adapt.bin
#100 67 (Frac 5)

#ResNet101

#pretrained=pretrained_models/cifar10_ADAM_001/ResNet101_avg_CIFAR-10_standard_best.bin
#100 54 (Frac 5)
#pretrained=pretrained_models/cifar10_ADAM_005/ResNet101_avg_CIFAR-10_standard_best.bin
#100 59 (Frac 5)
#pretrained=pretrained_models/cifar10_SGD_001/ResNet101_avg_CIFAR-10_standard_best.bin
#100 91 (Frac 5)

#pretrained=pretrained_models/cifar10_ADAM_001/ResNet101_avg_CIFAR-10_custom_best.bin
#does not exist
#pretrained=pretrained_models/cifar10_ADAM_005/ResNet101_avg_CIFAR-10_custom_best.bin
#100 79 (Frac 5)
#pretrained=pretrained_models/cifar10_SGD_001/ResNet101_avg_CIFAR-10_custom_best.bin
#100 94 (Frac 5)


#ResNet152

#pretrained=pretrained_models/old_models_hpmpc/ResNet152_cifar.bin
#does not match
#pretrained=pretrained_models/cifar10_ADAM_001/ResNet152_avg_CIFAR-10_standard_best.bin
#does not exist
#pretrained=pretrained_models/cifar10_ADAM_005/ResNet152_avg_CIFAR-10_standard_best.bin
#100 95 (Frac 5)
#pretrained=pretrained_models/cifar10_SGD_001/ResNet152_avg_CIFAR-10_standard_best.bin
#100 83 (Frac 5)

#pretrained=pretrained_models/cifar10_ADAM_001/ResNet152_avg_CIFAR-10_custom_best.bin
#does not exist
#pretrained=pretrained_models/cifar10_ADAM_005/ResNet152_avg_CIFAR-10_custom_best.bin
#100 93 (Frac 5)
#pretrained=pretrained_models/cifar10_SGD_001/ResNet152_avg_CIFAR-10_custom_best.bin
#100 95 (Frac 5)


#DRD

#pretrained=pretrained_models/old_models_hpmpc/model_DRD_C100_7K.bin
#pretrained=pretrained_models/old_models_hpmpc/model_DRD_C100_12K.bin
#pretrained=pretrained_models/old_models_hpmpc/model_DRD_C100_14K.bin
#pretrained=pretrained_models/old_models_hpmpc/model_DRD_C100_29K.bin
#pretrained=pretrained_models/old_models_hpmpc/model_DRD_C100_49K.bin
#pretrained=pretrained_models/old_models_hpmpc/model_DRD_C100_57K.bin
#pretrained=pretrained_models/old_models_hpmpc/model_DRD_C100_115K.bin
#pretrained=pretrained_models/old_models_hpmpc/model_DRD_C100_230K.bin
#pretrained=pretrained_models/old_models_hpmpc/model_DRD_C100_230K_mod.bin

# Uknown
#pretrained=pretrained_models/old_models_hpmpc/best_model.pth


#Images

#image_file=CIFAR-10_standard_test_images.bin
#label_file=CIFAR-10_standard_test_labels.bin
#image_file=CIFAR-10_custom_test_images.bin

#pretrained=AlexNet.bin
#pretrained=pretrained/vgg16_cifar_standard.bin
#pretrained=ResNet18_avg_CIFAR-10_standard_best.bin
#pretrained=ResNet18_avg.bin
#pretrained=LeNet_MNIST_standard_best.bin
#image_file=datasets/CIFAR-10_standard_test_images.bin
#label_file=datasets/CIFAR-10_standard_test_labels.bin
!pretrained=pretrained/Cifar_adam_005/ResNet18_avg_CIFAR-10_standard_best.bin
#pretrained=pretrained/Cifar_adam_005/AlexNet_CryptGPU_CIFAR-10_standard_best.bin
#image_file = datasets/CIFAR-10_custom_test_images.bin
#label_file = datasets/CIFAR-10_custom_test_labels.bin
#pretrained=AlexNetC.bin
!image_file=CIFAR-10_test_images.bin
!label_file=CIFAR-10_test_labels.bin
#pretrained=lenet5.bin
#image_file=MNIST_test_images.bin
#label_file=MNIST_test_labels.bin

#label_file=CIFAR-10_custom_test_labels.bin
#image_file=old/CIFAR-100_test_images.bin
#label_file=old/CIFAR-100_test_labels.bin
#image_file=CIFAR-100_standard_test_images.bin
#label_file=CIFAR-100_standard_test_labels.bin
#image_file=CIFAR-100_custom_test_images.bin
#label_file=CIFAR-100_custom_test_labels.bin

#image_file=MNIST_standard_test_images.bin
#label_file=MNIST_standard_test_labels.bin
#image_file=MNIST_custom_test_images.bin
#label_file=MNIST_custom_test_labels.bin




config.h
#pragma once

// === Basic Settings ===

#ifndef PROTOCOL
#define PROTOCOL 5
#endif

// Use a preprocessing phase? Only supported by Protocols 3,5,8,11,12
#ifndef PRE
#define PRE 0
#endif

// Party ID (starting from 0)
#ifndef PARTY
#define PARTY 2
#endif

// Bitlength of integers 
#ifndef BITLENGTH
#define BITLENGTH 32
#endif

// Fractional bits to use for fixed point arithmetic
#ifndef FRACTIONAL
#define FRACTIONAL 5
#endif

// 70+ Neural network architectures (LeNet, AlexNet, VGG, ResNet, etc.) on different dataset sizes (MNIST, CIFAR-10, Imagenet). Refer to programs/functions/NN.hpp
#ifndef FUNCTION_IDENTIFIER
#define FUNCTION_IDENTIFIER 1
#endif

// Number of inputs (mostly used by Benchmarking functions or Neural Networks)
#ifndef NUM_INPUTS
#define NUM_INPUTS 1
#endif


// === Concurrency Settings ===

// Register size to use for SIMD parallelization (Bitslicing/vectorization). Supported: 1,8,16,32,64,128(SSE),256(AVX-2),512(AVX-512)
// Info: MULT64 is supported by DATTYPE 64 and 512. MULT32 is supported for DATTYPE 32 and all DATATYPEs >= 128
#ifndef DATTYPE
#define DATTYPE 32
#endif

// Number of parallel processes to use
#ifndef PROCESS_NUM
#define PROCESS_NUM 1
#endif

// === Hardware Acceleration Settings ===

// 0 = xorshift, 1 = AES_BS, 2 = VAES/AES-NI. 0 is not secure.
#ifndef RANDOM_ALGORITHM
#define RANDOM_ALGORITHM 2
#endif

#ifndef USE_SSL_AES // USE SSl's AES implementation instead
#define USE_SSL_AES 0
#endif

#ifndef ARM
#define ARM 0 // 1 if ARM processor, 0 otherwise. Can speed up Sha hashing.
#endif


// USE CUDA for matrix multiplication?
#ifndef USE_CUDA_GEMM
#define USE_CUDA_GEMM 0
#endif


// === Tweaks ===

// How many gates should be buffered until sending them to the receiving party? 0 means the data of an entire communication round is buffered
#ifndef SEND_BUFFER
#define SEND_BUFFER 10000
#endif

// How many receiving messages should be buffered until the main thread is signaled that data is ready? 0 means that all data of a communication round needs to be ready before the main thread is signaled.
#ifndef RECV_BUFFER
#define RECV_BUFFER 10000
#endif

// How many messages should be buffered until a combined hash is performed? 0 means all hashes are calculated at the very end of the protocol.
#ifndef VERIFY_BUFFER
#define VERIFY_BUFFER 512/DATTYPE
#endif


// === Network Settings ===

// Use SSL encrypted communication?
#ifndef USE_SSL
#define USE_SSL 0
#endif


// === Neural Network Settings ===

#ifndef MODELOWNER
#define MODELOWNER P_0 // Who holds the model parameters? (-1: Dummy model parameters, P_0/P_1/P_2/P_3: Read locally from P_0/P_1/P_2/P_3 followed by secret sharing). Important: Use "P_0" not "0"!
#endif

#ifndef DATAOWNER
#define DATAOWNER P_1 // Who holds the data? (-1: Dummy dataset, P_0/P_1/P_2/P_3: Read locally from P_0/P_1/P_2/P_3 followed by secret sharing). Important: Use "P_0" not "0"!
#endif

#ifndef TRUNC_THEN_MULT
#define TRUNC_THEN_MULT 0 // 0: Truncate after multiplication, 1: Truncate before multiplication
#endif

#ifndef TRUNC_APPROACH
#define TRUNC_APPROACH 0 // 0: Probabilistic truncation, 1: Reduced Slack Truncation, 2: Exact Truncation
#endif

#ifndef TRUNC_DELAYED
#define TRUNC_DELAYED 0 // Delay CONV truncation until next ReLU 
#endif

#ifndef COMPUTE_ARGMAX
#define COMPUTE_ARGMAX 0 // 0: skip final argmax during inference, 1: Compute final argmax during inference
#endif

#ifndef PUBLIC_WEIGHTS
#define PUBLIC_WEIGHTS 0 // 0: weights are secretly shared, 1: weights are public
#endif


#ifndef COMPRESS
#define COMPRESS 0
#endif


// Reduced Bitlength that might be used for RELU, etc
#if COMPRESS == 0
#ifndef REDUCED_BITLENGTH_k
#define REDUCED_BITLENGTH_k BITLENGTH
#endif

#ifndef REDUCED_BITLENGTH_m
#define REDUCED_BITLENGTH_m 0
#endif

#else
#ifndef REDUCED_BITLENGTH_k
#define REDUCED_BITLENGTH_k 20
#endif

#ifndef REDUCED_BITLENGTH_m
#define REDUCED_BITLENGTH_m 12
#endif

#endif

#ifndef IS_TRAINING
#define IS_TRAINING 0 // Training or inference phase? Training is not supported yet.
#endif



// === Debugging Settings ===

// Print additional info?
#ifndef PRINT
#define PRINT 0
#endif

#ifndef PRINT_TIMINGS
#define PRINT_TIMINGS 1
#endif

#ifndef PRINT_IMPORTANT
#define PRINT_IMPORTANT 1
#endif


// === Other Settings ===


#ifndef SRNG_SEED
#define SRNG_SEED 0 // Seed for the random number generator.
#endif

// Starting port for required port range of the sockets, must be multiple of 1000 for some applications
#ifndef BASE_PORT
#define BASE_PORT 10000
#endif

#ifndef SPLIT_ROLES_OFFSET
#define SPLIT_ROLES_OFFSET 0
#endif

int base_port = BASE_PORT; // temporary solution

// Timeout in seconds when connecting to a socket
#ifndef CONNECTION_TIMEOUT
#define CONNECTION_TIMEOUT 500
#endif

// Timeout in milliseconds before attempting to connect again to a socket
#ifndef CONNECTION_RETRY
#define CONNECTION_RETRY 5
#endif


// === Legacy Settings ===

// Allow sharing of inputs in offline phase
#ifndef SHARE_PREP
#define SHARE_PREP 1
#endif


// Use optimized secret sharing? Often utilizes SRNG instead of secret sharing with communication
#ifndef OPT_SHARE
#define OPT_SHARE 1
#endif

// Use the initialization phase or import initialization data from a file?
#ifndef NO_INI
#define NO_INI 0
#endif

// Use the initialization phase or import initialization data from a file?
#ifndef INIT
#define INIT 1
#endif

// Use the online phase?
#ifndef LIVE
#define LIVE 1
#endif

// Use random inputs or inputs from a file? 
#ifndef INPUT
#define INPUT 'r'
#endif


// === Internal Settings, do not change ===

#if PROTOCOL == 4
#define HAS_POST_PROTOCOL 1
#elif (PROTOCOL == 3 || PROTOCOL == 5) && PARTY == 0
#define HAS_POST_PROTOCOL 1
#elif (PROTOCOL == 8 || PROTOCOL == 11 || PROTOCOL == 12) && PARTY == 3
#define HAS_POST_PROTOCOL 1
#endif

#if PROTOCOL == 4
#ifndef num_players
#define num_players 2
#endif
#elif PROTOCOL < 7
#ifndef num_players
#define num_players 3
#endif
#else
#ifndef num_players
#define num_players 4
#endif
#endif

#if PROTOCOL > 7
#ifndef MAL
#define MAL 1
#endif
#endif

#if num_players == 2
#define PNEXT 0
#define PPREV 0
#define PSELF 1
#if PARTY == 0
#define P_0 1
#define P_1 0
#elif PARTY == 1
#define P_0 0
#define P_1 1
#endif
#elif num_players == 3
    #define PSELF 2
    #if PARTY == 0
        #define P_0 2
        #define P_1 0
        #define P_2 1
        #define PPREV 1
        #define PNEXT 0
    #elif PARTY == 1
        #define P_0 0
        #define P_1 2
        #define P_2 1
        #define PPREV 0
        #define PNEXT 1
    #elif PARTY == 2
        #define P_0 0
        #define P_1 1
        #define P_2 2
        #define PPREV 1
        #define PNEXT 0
    #endif
#elif num_players == 4
    #define PSELF 3
    #define P_0123 3
    #define P_012 4
    #define P_013 5
    #define P_023 6
    #define P_123 7
    #define P_123_2 3 // Trick for Protocols 10-12

    #if PARTY == 0
        #define P_0 3
        #define P_1 0
        #define P_2 1
        #define P_3 2
        #define PPREV 2
        #define PNEXT 0
        #define PMIDDLE 1
    #elif PARTY == 1
        #define P_0 0
        #define P_1 3
        #define P_2 1
        #define P_3 2
        #define PPREV 0
        #define PNEXT 1
        #define PMIDDLE 2
    #elif PARTY == 2
        #define P_0 0
        #define P_1 1
        #define P_2 3
        #define P_3 2
        #define PPREV 1
        #define PNEXT 2
        #define PMIDDLE 0
    #elif PARTY == 3
        #define P_0 0
        #define P_1 1
        #define P_2 2
        #define P_3 3
        #define PPREV 2
        #define PNEXT 0
        #define PMIDDLE 1
    #endif
#endif

#if FUNCTION_IDENTIFIER > 5
#ifndef MULTI_INPUT
#define MULTI_INPUT 1
#endif
#else
#ifndef MULTI_INPUT
#define MULTI_INPUT 0
#endif
#endif
/* #endif */

#if FUNCTION_IDENTIFIER < 65
#if FUNCTION_IDENTIFIER == 46 || FUNCTION_IDENTIFIER == 49 || FUNCTION_IDENTIFIER == 52 || FUNCTION_IDENTIFIER == 59 || FUNCTION_IDENTIFIER == 62 // RCA
#ifndef BANDWIDTH_OPTIMIZED
#define BANDWIDTH_OPTIMIZED 1
#endif

#ifndef ONLINE_OPTIMIZED
#define ONLINE_OPTIMIZED 0
#endif
#elif FUNCTION_IDENTIFIER == 47 || FUNCTION_IDENTIFIER == 50 || FUNCTION_IDENTIFIER == 53 || FUNCTION_IDENTIFIER == 60 || FUNCTION_IDENTIFIER == 63 // PPA
#ifndef BANDWIDTH_OPTIMIZED
#define BANDWIDTH_OPTIMIZED 0
#endif

#ifndef ONLINE_OPTIMIZED
#define ONLINE_OPTIMIZED 0
#endif
#elif FUNCTION_IDENTIFIER == 48 || FUNCTION_IDENTIFIER == 51 || FUNCTION_IDENTIFIER == 54 || FUNCTION_IDENTIFIER == 61 || FUNCTION_IDENTIFIER == 64 // PPA 4-Way
#ifndef BANDWIDTH_OPTIMIZED
#define BANDWIDTH_OPTIMIZED 0
#endif

#ifndef ONLINE_OPTIMIZED
#define ONLINE_OPTIMIZED 1
#endif
#endif
#elif FUNCTION_IDENTIFIER < 400
#if FUNCTION_IDENTIFIER < 100  // RCA
#ifndef BANDWIDTH_OPTIMIZED
#define BANDWIDTH_OPTIMIZED 1 // 1 if bandwidth optimized (e.g. Ripple Carry Adder), 0 if Latency optimized (e.g. Multi-input AND gates, Parallel Prefix Adder)
#endif

#ifndef ONLINE_OPTIMIZED
#define ONLINE_OPTIMIZED 0 // 1 if online optimized (e.g. MULTI_INPUT AND gates), 0 if optimized for total communication (e.g. no MULTI_INPUT AND gates)
#endif
#elif FUNCTION_IDENTIFIER < 200 // PPA
#ifndef BANDWIDTH_OPTIMIZED
#define BANDWIDTH_OPTIMIZED 0
#endif

#ifndef ONLINE_OPTIMIZED
#define ONLINE_OPTIMIZED 0
#endif
#elif FUNCTION_IDENTIFIER < 300
#ifndef BANDWIDTH_OPTIMIZED
#define BANDWIDTH_OPTIMIZED 0 // PPA 4-Way
#endif

#ifndef ONLINE_OPTIMIZED
#define ONLINE_OPTIMIZED 1
#endif
#endif
#else
#if FUNCTION_IDENTIFIER == 404 || FUNCTION_IDENTIFIER == 407 || FUNCTION_IDENTIFIER == 412 // RCA
#ifndef BANDWIDTH_OPTIMIZED
#define BANDWIDTH_OPTIMIZED 1
#endif

#ifndef ONLINE_OPTIMIZED
#define ONLINE_OPTIMIZED 0
#endif
#elif FUNCTION_IDENTIFIER == 405 || FUNCTION_IDENTIFIER == 408 || FUNCTION_IDENTIFIER == 413 // PPA
#ifndef BANDWIDTH_OPTIMIZED
#define BANDWIDTH_OPTIMIZED 0
#endif

#ifndef ONLINE_OPTIMIZED
#define ONLINE_OPTIMIZED 0
#endif
#elif FUNCTION_IDENTIFIER == 406 || FUNCTION_IDENTIFIER == 409 || FUNCTION_IDENTIFIER == 414 // PPA 4-Way
#ifndef BANDWIDTH_OPTIMIZED
#define BANDWIDTH_OPTIMIZED 0
#endif

#ifndef ONLINE_OPTIMIZED
#define ONLINE_OPTIMIZED 1
#endif
#endif
#endif


#ifndef SIMULATE_MPC_FUNCTIONS
#define SIMULATE_MPC_FUNCTIONS 0
#endif

#if BITLENGTH == 64
#ifndef INT_TYPE
#define INT_TYPE int64_t
#endif

#ifndef UINT_TYPE
#define UINT_TYPE uint64_t
#endif

#ifndef LOG2_BITLENGTH
#define LOG2_BITLENGTH 6
#endif

#ifndef LOG4_BITLENGTH
#define LOG4_BITLENGTH 3
#endif
#elif BITLENGTH == 32
#ifndef INT_TYPE
#define INT_TYPE int32_t
#endif

#ifndef UINT_TYPE
#define UINT_TYPE uint32_t
#endif

#ifndef LOG2_BITLENGTH
#define LOG2_BITLENGTH 5
#endif

#ifndef LOG4_BITLENGTH
#define LOG4_BITLENGTH 3
#endif
#elif BITLENGTH == 16
#ifndef INT_TYPE
#define INT_TYPE int16_t
#endif

#ifndef UINT_TYPE
#define UINT_TYPE uint16_t
#endif

#ifndef LOG2_BITLENGTH
#define LOG2_BITLENGTH 4
#endif

#ifndef LOG4_BITLENGTH
#define LOG4_BITLENGTH 2
#endif
#elif BITLENGTH == 8
#ifndef INT_TYPE
#define INT_TYPE int8_t
#endif

#ifndef UINT_TYPE
#define UINT_TYPE uint8_t
#endif

#ifndef LOG2_BITLENGTH
#define LOG2_BITLENGTH 3
#endif

#ifndef LOG4_BITLENGTH
#define LOG4_BITLENGTH 2
#endif
#endif


#ifndef JIT_VEC
#define JIT_VEC 1 // 0: vectorize and share inputs from the beginning, 1: vectorize and share inputs just in time, load a batch of images, then vectorize
#endif

#ifndef BASETYPE
#define BASETYPE 0 // 0: Additive_Share, 1: sint
#endif

#if JIT_VEC == 0
#ifndef BASE_DIV
#define BASE_DIV 1
#endif
#else
#if BASETYPE == 0
#ifndef BASE_DIV
#define BASE_DIV DATTYPE/BITLENGTH
#endif
#else
#ifndef BASE_DIV
#define BASE_DIV DATTYPE
#endif
#endif
#endif


#ifndef PHASE_LIVE
#define PHASE_LIVE 1
#endif

#ifndef PHASE_INIT
#define PHASE_INIT 0
#endif

#ifndef PHASE_PRE
#define PHASE_PRE 2
#endif

